src/components/DateRangeInput/index.tsx

```diff
@@ -0,0 +1,370 @@
+    inputClassName?: string;
+    value?: Value;
+    name: N;
+    onChange?: (value: Value | undefined, name: N) => void;
+}
+
+function DateRangeInput<N extends string | number | undefined>(props: Props<N>) {
+    const {
+        actions,
+        actionsContainerClassName,
+        className,
+        disabled,
+        error,
+        errorContainerClassName,
+        hint,
+        hintContainerClassName,
+        icons,
+        iconsContainerClassName,
+        inputSectionClassName,
+        label,
+        labelContainerClassName,
+        readOnly,
+        uiMode,
+        inputElementRef,
+        containerRef: containerRefFromProps,
+        inputSectionRef,
+        inputClassName,
+        onChange,
+        name,
+        value,
+    } = props;
+
+    const prevMonthDate = new Date();
+    prevMonthDate.setMonth(prevMonthDate.getMonth() - 1);
+    const prevMonthDateString = ymdToDateString(
+        prevMonthDate.getFullYear(),
+        prevMonthDate.getMonth(),
+        1,
+    );
+
+    const [tempStartDate, setTempStartDate] = React.useState<string | undefined>();
+    const [calendarMonthSelectionPopupClassName] = React.useState(randomString(16));
+    const createdContainerRef = React.useRef<HTMLDivElement>(null);
+    const popupRef = React.useRef<HTMLDivElement>(null);
+
+    const containerRef = containerRefFromProps ?? createdContainerRef;
+    const uiModeClassName = useUiModeClassName(uiMode, styles.light, styles.dark);
+    const [
+        showCalendar,
+        setShowCalendarTrue,
+        setShowCalendarFalse,,
+        toggleShowCalendar,
+    ] = useBooleanState(false);
+
+    const hideCalendar = React.useCallback(() => {
+        setTempStartDate(undefined);
+        setShowCalendarFalse();
+    }, [setShowCalendarFalse]);
+
+    const handlePopupBlur = React.useCallback(
+        (isClickedWithin: boolean, e: MouseEvent) => {
+            // Following is to prevent the popup blur when
+            // month selection is changed in the calendar
+            const container = document.getElementsByClassName(
+                calendarMonthSelectionPopupClassName,
+            )[0];
+            const isContainerOrInsideContainer = container
+                ? container === e.target || container.contains(e.target as HTMLElement)
+                : false;
+            if (!isClickedWithin && !isContainerOrInsideContainer) {
+                hideCalendar();
+            }
+        },
+        [hideCalendar, calendarMonthSelectionPopupClassName],
+    );
+
+    useBlurEffect(
+        showCalendar,
+        handlePopupBlur,
+        popupRef,
+        containerRef,
+    );
+
+    const dateRendererParams = React.useCallback(() => ({
+        startDate: tempStartDate ?? value?.startDate,
+        endDate: !tempStartDate ? value?.endDate : undefined,
+    }), [tempStartDate, value]);
+
+    const handleCalendarDateClick: CalendarProps<never>['onDateClick'] = React.useCallback(
```

Comment:
This function is complicated. It does a lot of things.
Let's try to break this down.